File: ios/runanywhere-sdks-latest/examples/ios/RunAnywhereAI/RunAnywhereAI/Features/Benchmarks/ViewModels/BenchmarkViewModel.swift
//
//  BenchmarkViewModel.swift
//  RunAnywhereAI
//
//  Orchestrates benchmark execution, persistence, and export.
//

import Foundation
#if canImport(UIKit)
import UIKit
#endif

@MainActor
@Observable
final class BenchmarkViewModel {

    // MARK: - State

    var isRunning = false
    var progress: Double = 0
    var currentScenario: String = ""
    var currentModel: String = ""
    var completedCount: Int = 0
    var totalCount: Int = 0
    var pastRuns: [BenchmarkRun] = []
    var selectedCategories: Set<BenchmarkCategory> = Set(BenchmarkCategory.allCases)
    var errorMessage: String?
    var showClearConfirmation = false

    /// Brief toast message shown after clipboard copy
    var copiedToastMessage: String?

    /// Categories that had no downloaded models during the last run
    var skippedCategoriesMessage: String?

    // MARK: - Private

    private let runner = BenchmarkRunner()
    private let store = BenchmarkStore()
    private var runTask: Task<Void, Never>?
    private var toastTask: Task<Void, Never>?

    // MARK: - Lifecycle

    func loadPastRuns() {
        pastRuns = store.loadRuns().reversed()
    }

    // MARK: - Run

    func runBenchmarks() {
        guard !isRunning else { return }
        isRunning = true
        errorMessage = nil
        skippedCategoriesMessage = nil
        progress = 0
        completedCount = 0
        totalCount = 0
        currentScenario = "Preparing..."
        currentModel = ""

        runTask = Task {
            let deviceInfo = makeDeviceInfo()
            var run = BenchmarkRun(deviceInfo: deviceInfo)

            do {
                let output = try await runner.runBenchmarks(
                    categories: selectedCategories
                ) { [weak self] update in
                    Task { @MainActor in
                        self?.progress = update.progress
                        self?.completedCount = update.completedCount
                        self?.totalCount = update.totalCount
                        self?.currentScenario = update.currentScenario
                        self?.currentModel = update.currentModel
                    }
                }

                if !output.skippedCategories.isEmpty {
                    let names = output.skippedCategories.map(\.displayName).joined(separator: ", ")
                    skippedCategoriesMessage = "Skipped (no models): \(names)"
                }

                run.results = output.results
                run.status = output.results.allSatisfy(\.metrics.didSucceed) ? .completed : .failed
                run.completedAt = Date()
            } catch is CancellationError {
                run.status = .cancelled
                run.completedAt = Date()
            } catch let error as BenchmarkRunnerError {
                run.status = .failed
                run.completedAt = Date()
                errorMessage = error.localizedDescription
            } catch {
                run.status = .failed
                run.completedAt = Date()
                errorMessage = error.localizedDescription
            }

            // Only save if we got results or it was explicitly cancelled/failed
            if !run.results.isEmpty || run.status != .running {
                store.save(run: run)
            }
            loadPastRuns()
            isRunning = false
        }
    }

    func cancel() {
        runTask?.cancel()
        runTask = nil
    }

    func clearAllResults() {
        store.clearAll()
        pastRuns = []
    }

    // MARK: - Copy to Clipboard

    func copyToClipboard(run: BenchmarkRun, format: BenchmarkExportFormat) {
        #if canImport(UIKit)
        let report = BenchmarkReportFormatter.formattedString(run: run, format: format)
        UIPasteboard.general.string = report
        let generator = UINotificationFeedbackGenerator()
        generator.notificationOccurred(.success)
        copiedToastMessage = "\(format.displayName) copied!"
        // Auto-dismiss after 2s — cancel any previous dismiss task
        toastTask?.cancel()
        toastTask = Task {
            try? await Task.sleep(for: .seconds(2))
            copiedToastMessage = nil
        }
        #endif
    }

    // MARK: - File Export

    func shareJSON(run: BenchmarkRun) -> URL {
        BenchmarkReportFormatter.writeJSON(run: run)
    }

    func shareCSV(run: BenchmarkRun) -> URL {
        BenchmarkReportFormatter.writeCSV(run: run)
    }

    // MARK: - Helpers

    private func makeDeviceInfo() -> BenchmarkDeviceInfo {
        if let sysInfo = DeviceInfoService.shared.deviceInfo {
            return BenchmarkDeviceInfo.fromSystem(sysInfo)
        }
        return BenchmarkDeviceInfo(
            modelName: "Unknown",
            chipName: "Unknown",
            totalMemoryBytes: Int64(ProcessInfo.processInfo.physicalMemory),
            availableMemoryBytes: SyntheticInputGenerator.availableMemoryBytes(),
            osVersion: ProcessInfo.processInfo.operatingSystemVersionString
        )
    }
}

File: ios/runanywhere-sdks-latest/examples/android/RunAnywhereAI/app/src/main/java/com/runanywhere/runanywhereai/presentation/benchmarks/viewmodel/BenchmarkViewModel.kt
package com.runanywhere.runanywhereai.presentation.benchmarks.viewmodel

import android.app.Application
import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import android.content.Intent
import androidx.core.content.FileProvider
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.runanywhere.runanywhereai.presentation.benchmarks.models.BenchmarkCategory
import com.runanywhere.runanywhereai.presentation.benchmarks.models.BenchmarkDeviceInfo
import com.runanywhere.runanywhereai.presentation.benchmarks.models.BenchmarkRun
import com.runanywhere.runanywhereai.presentation.benchmarks.models.BenchmarkRunStatus
import com.runanywhere.runanywhereai.presentation.benchmarks.models.BenchmarkStore
import com.runanywhere.runanywhereai.presentation.benchmarks.services.BenchmarkRunner
import com.runanywhere.runanywhereai.presentation.benchmarks.services.BenchmarkRunnerError
import com.runanywhere.runanywhereai.presentation.benchmarks.utilities.BenchmarkExportFormat
import com.runanywhere.runanywhereai.presentation.benchmarks.utilities.BenchmarkReportFormatter
import com.runanywhere.runanywhereai.presentation.benchmarks.utilities.SyntheticInputGenerator
import com.runanywhere.sdk.models.DeviceInfo
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import java.io.File

// -- UI State --

data class BenchmarkUiState(
    val isRunning: Boolean = false,
    val progress: Float = 0f,
    val currentScenario: String = "",
    val currentModel: String = "",
    val completedCount: Int = 0,
    val totalCount: Int = 0,
    val pastRuns: List<BenchmarkRun> = emptyList(),
    val selectedCategories: Set<BenchmarkCategory> = BenchmarkCategory.entries.toSet(),
    val errorMessage: String? = null,
    val showClearConfirmation: Boolean = false,
    val copiedToastMessage: String? = null,
    val skippedCategoriesMessage: String? = null,
)

/**
 * Orchestrates benchmark execution, persistence, and export.
 * Matches iOS BenchmarkViewModel exactly.
 */
class BenchmarkViewModel(application: Application) : AndroidViewModel(application) {

    private val _uiState = MutableStateFlow(BenchmarkUiState())
    val uiState: StateFlow<BenchmarkUiState> = _uiState.asStateFlow()

    private val runner = BenchmarkRunner()
    private val store = BenchmarkStore(application)
    private var runJob: Job? = null

    init {
        loadPastRuns()
    }

    // -- Lifecycle --

    fun loadPastRuns() {
        _uiState.update { it.copy(pastRuns = store.loadRuns().reversed()) }
    }

    // -- Category Toggle --

    fun toggleCategory(category: BenchmarkCategory) {
        _uiState.update { state ->
            val updated = state.selectedCategories.toMutableSet()
            if (category in updated) updated.remove(category) else updated.add(category)
            state.copy(selectedCategories = updated)
        }
    }

    fun selectAllCategories() {
        _uiState.update { it.copy(selectedCategories = BenchmarkCategory.entries.toSet()) }
    }

    // -- Run --

    fun runBenchmarks() {
        if (_uiState.value.isRunning) return

        _uiState.update {
            it.copy(
                isRunning = true,
                errorMessage = null,
                skippedCategoriesMessage = null,
                progress = 0f,
                completedCount = 0,
                totalCount = 0,
                currentScenario = "Preparing...",
                currentModel = "",
            )
        }

        runJob = viewModelScope.launch {
            val deviceInfo = makeDeviceInfo()
            var run = BenchmarkRun(deviceInfo = deviceInfo)

            try {
                val runResult = runner.runBenchmarks(
                    categories = _uiState.value.selectedCategories,
                ) { update ->
                    _uiState.update { state ->
                        state.copy(
                            progress = update.progress,
                            completedCount = update.completedCount,
                            totalCount = update.totalCount,
                            currentScenario = update.currentScenario,
                            currentModel = update.currentModel,
                        )
                    }
                }

                if (runResult.skippedCategories.isNotEmpty()) {
                    val names = runResult.skippedCategories.joinToString { it.displayName }
                    _uiState.update { it.copy(skippedCategoriesMessage = "Skipped (no models): $names") }
                }

                run = run.copy(
                    results = runResult.results,
                    status = BenchmarkRunStatus.COMPLETED,
                    completedAt = System.currentTimeMillis(),
                )
            } catch (_: kotlinx.coroutines.CancellationException) {
                run = run.copy(
                    status = BenchmarkRunStatus.CANCELLED,
                    completedAt = System.currentTimeMillis(),
                )
            } catch (e: BenchmarkRunnerError) {
                run = run.copy(
                    status = BenchmarkRunStatus.FAILED,
                    completedAt = System.currentTimeMillis(),
                )
                _uiState.update { it.copy(errorMessage = e.message) }
            } catch (e: Exception) {
                run = run.copy(
                    status = BenchmarkRunStatus.FAILED,
                    completedAt = System.currentTimeMillis(),
                )
                _uiState.update { it.copy(errorMessage = e.localizedMessage ?: e.message ?: "Unknown error") }
            }

            if (run.results.isNotEmpty() || run.status != BenchmarkRunStatus.RUNNING) {
                store.save(run)
            }
            loadPastRuns()
            _uiState.update { it.copy(isRunning = false) }
        }
    }

    fun cancel() {
        runJob?.cancel()
        runJob = null
    }

    // -- Clear --

    fun showClearConfirmation() {
        _uiState.update { it.copy(showClearConfirmation = true) }
    }

    fun dismissClearConfirmation() {
        _uiState.update { it.copy(showClearConfirmation = false) }
    }

    fun clearAllResults() {
        store.clearAll()
        _uiState.update { it.copy(pastRuns = emptyList(), showClearConfirmation = false) }
    }

    fun dismissError() {
        _uiState.update { it.copy(errorMessage = null) }
    }

    // -- Copy to Clipboard --

    fun copyToClipboard(run: BenchmarkRun, format: BenchmarkExportFormat) {
        val context = getApplication<Application>()
        val report = BenchmarkReportFormatter.formattedString(run, format, context)
        val clipboard = context.getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
        clipboard.setPrimaryClip(ClipData.newPlainText("Benchmark Report", report))
        _uiState.update { it.copy(copiedToastMessage = "${format.displayName} copied!") }
        viewModelScope.launch {
            delay(2000)
            _uiState.update { it.copy(copiedToastMessage = null) }
        }
    }

    // -- File Export --

    fun shareFile(run: BenchmarkRun, csv: Boolean): Intent {
        val context = getApplication<Application>()
        val file: File = if (csv) {
            BenchmarkReportFormatter.writeCSV(run, context)
        } else {
            BenchmarkReportFormatter.writeJSON(run, context)
        }
        val uri = FileProvider.getUriForFile(
            context,
            "${context.packageName}.fileprovider",
            file,
        )
        return Intent(Intent.ACTION_SEND).apply {
            type = if (csv) "text/csv" else "application/json"
            putExtra(Intent.EXTRA_STREAM, uri)
            addFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION)
        }
    }

    // -- Helpers --

    private fun makeDeviceInfo(): BenchmarkDeviceInfo {
        return try {
            val info = DeviceInfo.current
            BenchmarkDeviceInfo(
                modelName = info.modelName,
                chipName = info.architecture,
                totalMemoryBytes = info.totalMemory,
                availableMemoryBytes = SyntheticInputGenerator.availableMemoryBytes(),
                osVersion = "Android ${info.osVersion}",
            )
        } catch (_: Exception) {
            BenchmarkDeviceInfo(
                modelName = android.os.Build.MODEL,
                chipName = android.os.Build.HARDWARE,
                totalMemoryBytes = Runtime.getRuntime().maxMemory(),
                availableMemoryBytes = SyntheticInputGenerator.availableMemoryBytes(),
                osVersion = "Android ${android.os.Build.VERSION.RELEASE}",
            )
        }
    }
}
